Tile highlighting hexxle 126

RicardoMonteiroSimoes opened the pull request.

DennisBannerman commented:
Testing it right now. Things I noticed (will edit, if I find more):

* If you open the pausemenu, you can still click behind it, tiles are also still highlighted. This functionality was disabled in a previous commit. Maybe change it back/reimplement unless that is intended behaviour.

RicardoMonteiroSimoes commented:
> Testing it right now. Things I noticed (will edit, if I find more):
> 
> * If you open the pausemenu, you can still click behind it, tiles are also still highlighted. This functionality was disabled in a previous commit. Maybe change it back/reimplement unless that is intended behaviour.
> * I seem to have Recovery GameObjects in my main scene - is that only the case on my machine or do you have those too?

I deleted those objects, I assume some relics I must've overlooked after the rebase.

I fixed those bugs and added tests to cover it, should be okay now

DennisBannerman commented on unity/Assets/Scripts/Unity/Input/MouseEventsHandler.cs:
```diff
@@ -0,0 +1,102 @@
+        void Start()
+        {
+            var inputManager = new InputManager();
+            inputManager.TilePlacement.MouseClick.Enable();
+            inputManager.TilePlacement.MouseClick.performed += context => MouseClickAction();
```
I'd move this to OnEnable() and also make sure, that OnDisable() is implemented and strips `performed` of its Eventhandler method. See EscapeEventHandler class.

DennisBannerman commented on unity/Assets/Scripts/Unity/Input/MouseEventsHandler.cs:
```diff
@@ -0,0 +1,102 @@
+            }
+        }
+
+        public void PauseGame()
+        {
+           isGamePaused = true;
+           if(currentCollisionTile != null)
+           {
+                currentCollisionTile.GetComponent<UnityTileHighlighter>().TurnOff();
+           }
+            
+        }
+
+        public void ResumeGame()
+        {
+            isGamePaused = false;
+            if (currentCollisionTile != null)
+            {
+                currentCollisionTile.GetComponent<UnityTileHighlighter>().TurnOn();
+            }
+        }
```
Does UnityTileHighlighter's `enabled` property not already provide the TurnOn/TurnOff logic? Setting enabled to true/false should fire `OnDisable` and `OnEnable`, which you chould also override if you'd need some custom behaviour.

RicardoMonteiroSimoes replied:
no idea, didn't think of that before. I'll try it out

DennisBannerman commented on unity/Assets/Scripts/Unity/UnityStack.cs:
```diff
@@ -56,7 +56,7 @@ private void DisplayStack()
                 ITile tile = firstTen.ElementAt(i);
                 Texture texture = GetTextureForTileType(tile);
-                GameObject newTile = Instantiate(Template, Content.transform);
+                GameObject newTile = Instantiate(Template);
```
What does this change? Just curious.

RicardoMonteiroSimoes replied:
I had a warning telling me there are too many transforms going on, removing this parameter fixed it.

DennisBannerman commented on unity/Assets/Scripts/Unity/UnityTileHighlighter.cs:
```diff
@@ -0,0 +1,59 @@
+            isHighlighted = false;
+        }
+
+        // Update is called once per frame
+        void Update()
+        {
+
+        }
```
Looks like unused code.

DennisBannerman commented on unity/Assets/Scripts/Unity/UnityTileHighlighter.cs:
```diff
@@ -0,0 +1,59 @@
+            }
+        }
+
+        public void TurnOn()
+        {
+            isHighlighted = true;
+            setToBlackOnce = false;
+        }
+
+        public void TurnOff()
+        {
+            isHighlighted = false;
+            setToBlackOnce = true;
+        }
```
As noted in another class - is this required if we can just enable or disable the component?

DennisBannerman reviewed (approved):
Looks good, got some minor changes. They don't change the basic logic though so feel free to merge at your own discretion.

RicardoMonteiroSimoes commented:
Refactored the code to use `enabled` to turn on and off, works exactly the same but with much cleaner code
